### PR TITLE
check_repeat_nonrepeat Function Optimization

### PR DESCRIPTION
**Pull request by rsh52**
# Description

## Background
The `check_repeat_nonrepeat` function is used to identify edge case scenarios in longitudinal databases (those containing event arms) where REDCap users may make an instrument repeating _**and**_ nonrepeating in different events. This is a behavior that is not recommended and `REDCapTidieR` should not support. 

An example export would look something like below:

|record_id|redcap_event_name|redcap_repeat_instrument|redcap_repeat_instance|combination_variable|
|---------|-----------------|------------------------|----------------------|--------------------|
|1        |event_1          |NA                      |NA                    |A                   |
|2        |event_2          |combination             |1                     |B                   |
|3        |event_3          |combination             |2                     |C                   |

Where `combination_variable` contains data for rows where `redcap_repeat_*` is both `NA` and populated.

## Proposed Solution

When tested againt the Prodigy database, the previous method involved nested loops across all variables and their rows to check this behavior, resulting in `profvis` performance of 1740ms:
image

The new method creates dummy variables for each column and assigns values of "repeat" or "nonrepeat" per row. Then a `map`ping function determines if a dummy column has both "repeat" and "nonrepeat" identifiers and errors to return the name of the first column offender. The new performance has improved to 10ms:
image

# Proposed Changes 
List changes below in bullet format:

- Update the `check_repeat_nonrepeat` function
- Update `roxygen` documentation accordingly

### Issue Addressed
**Addresses some of the performance issues mentioned in #48, but more improvements are necessary.**

## Future Work

The performance improvement of `read_redcap_tidy` as a whole against the test Prodigy database improved from 23710ms to 18530ms, so there is still more improvement to go.

While the current method is better than the previous one, it may not be desirable to create a dataframe twice as large as the original database output. This could potentially be unappealing in exceptionally large database extractions.

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [NA] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

### CODE REVIEW

File: R/checks.R [hunk 1]
```diff
@@ -57,51 +58,60 @@ check_user_rights <- function(
...
+  #    analysis (i.e. "safe columns").
+  # 2) Create a dummy column for each remaining column and use case_when to
+  #    to assign whether the column demonstrates repeating, nonrepeating, or
+  #    indeterminant behavior per row. Indetemrinant would be the case for
+  #    data not yet filled out.
+  # 3) Use a mapping function to determine if any dummy columns contain both
+  #    "repeating" AND "nonrepeating" declarations, if so error out.
+
+  # Step (1)
+  safe_cols <- names(db_data)[1:4]
```
Comment (skadauke) [inline R/checks.R:81]: Is this true that the first 4 columns of `db_data` are always safe? What if the database has arms or no arms?

File: R/checks.R [hunk 2]
```diff
@@ -57,51 +58,60 @@ check_user_rights <- function(
...
+    if ("repeating" %in% check_data &
+        "nonrepeating" %in% check_data){
       stop(
-        paste0("Instrument detected belonging to an instrument that is both repeating and nonrepeating: ",
-               db_data %>%
-                 select(.data$redcap_repeat_instrument, names(db_data)[i]) %>%
-                 drop_na(everything()) %>%
-                 slice(1) %>%
-                 pull(.data$redcap_repeat_instrument)
+        paste0("Instrument detected belonging to an instrument that is both
```
Comment (skadauke) [inline R/checks.R:102]: Can you revise that phrase?

**Review by skadauke — CHANGES_REQUESTED** [review]

File: R/checks.R (hunk 2 above)
Comment (rsh52) [reply R/checks.R:102]: How about:
`"Instrument detected with an assigned structure of both repeating and nonrepeating: "`

**Comment by skadauke** [comment]
Yeah I think this is better! How about:

"Instrument detected that has both repeated and nonrepeated instances defined in the project: "
________________________________
From: Rich Hanna ***@***.***>
Sent: Tuesday, October 18, 2022 3:50 PM
To: CHOP-CGTDataOps/REDCapTidieR ***@***.***>
Cc: Stephan Kadauke ***@***.***>; Review requested ***@***.***>
Subject: [External]Re: [CHOP-CGTDataOps/REDCapTidieR] check_repeat_nonrepeat Function Optimization (PR #49)


@rsh52 commented on this pull request.

________________________________

In R/checks.R<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTDataOps%2FREDCapTidieR%2Fpull%2F49%23discussion_r998645336&data=05%7C01%7Ckadaukes%40chop.edu%7C5310b421f19f456d217308dab14213d6%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017194656872601%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Zia9Un62hnHEebsCcgM66%2FRS6iPALjofACs3xAGCO00%3D&reserved=0>:

>        stop(
-        paste0("Instrument detected belonging to an instrument that is both repeating and nonrepeating: ",
-               db_data %>%
-                 select(.data$redcap_repeat_instrument, names(db_data)[i]) %>%
-                 drop_na(everything()) %>%
-                 slice(1) %>%
-                 pull(.data$redcap_repeat_instrument)
+        paste0("Instrument detected belonging to an instrument that is both


How about:
"Instrument detected with an assigned structure of both repeating and nonrepeating: "

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTDataOps%2FREDCapTidieR%2Fpull%2F49%23discussion_r998645336&data=05%7C01%7Ckadaukes%40chop.edu%7C5310b421f19f456d217308dab14213d6%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017194657028835%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=X2WkpfRzbDSQtcdA4FS1c9aBEC0lqP%2FKSc44ZapaFu4%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTGHWU7ZHHA6Z2QWU7HHQTWD35SBANCNFSM6AAAAAARILIAO4&data=05%7C01%7Ckadaukes%40chop.edu%7C5310b421f19f456d217308dab14213d6%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017194657028835%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=zBALYTbCHn3R6Ac1bVEx1MD6NTFFnHwGjQEWRptuYG0%3D&reserved=0>.
You are receiving this because your review was requested.Message ID: ***@***.***>

** This email originated from an EXTERNAL sender to CHOP. Proceed with caution when replying, opening attachments, or clicking links. Do not disclose your CHOP credentials, employee information, or protected health information to a potential hacker**.

File: R/checks.R (hunk 1 above)
Comment (rsh52) [reply R/checks.R:81]: Ah that's a good point, I think I can correct this to look for the first 3 instead when `redcap_event_name` is missing i.e. `is_longitudinal` is T/F and reorder things.

**Comment by skadauke** [comment]
I think it's OK to say column #1 is safe since it's always the record ID and name can vary. However the additional columns I would enumerate by name.
________________________________
From: Rich Hanna ***@***.***>
Sent: Tuesday, October 18, 2022 3:54 PM
To: CHOP-CGTDataOps/REDCapTidieR ***@***.***>
Cc: Stephan Kadauke ***@***.***>; Review requested ***@***.***>
Subject: [External]Re: [CHOP-CGTDataOps/REDCapTidieR] check_repeat_nonrepeat Function Optimization (PR #49)


@rsh52 commented on this pull request.

________________________________

In R/checks.R<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTDataOps%2FREDCapTidieR%2Fpull%2F49%23discussion_r998647951&data=05%7C01%7Ckadaukes%40chop.edu%7Cb1c7e775ecd44062e36d08dab1428e9e%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017196660970123%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=jUr3L2OChXmEnBXQBh5DEzVqTolSuCQWzcnxxX9UgRc%3D&reserved=0>:

> -    repeating <- NA
-
-    for (j in seq_len(nrow(db_data))){
+  # This check function looks for potential repeat/nonrepeat behavior using the
+  # steps below:
+  # 1) Define standard columns that don't need checking and remove those from
+  #    analysis (i.e. "safe columns").
+  # 2) Create a dummy column for each remaining column and use case_when to
+  #    to assign whether the column demonstrates repeating, nonrepeating, or
+  #    indeterminant behavior per row. Indetemrinant would be the case for
+  #    data not yet filled out.
+  # 3) Use a mapping function to determine if any dummy columns contain both
+  #    "repeating" AND "nonrepeating" declarations, if so error out.
+
+  # Step (1)
+  safe_cols <- names(db_data)[1:4]


Ah that's a good point, I think I can correct this to look for the first 3 instead when redcap_event_name is missing i.e. is_longitudinal is T/F and reorder things.

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTDataOps%2FREDCapTidieR%2Fpull%2F49%23discussion_r998647951&data=05%7C01%7Ckadaukes%40chop.edu%7Cb1c7e775ecd44062e36d08dab1428e9e%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017196661126347%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=F%2FI4KLATylqcYtjcTNh8ZHO7SaFEWSkgIpJgePFWJzo%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTGHWUK3PJUGJYGZE6QE2LWD3565ANCNFSM6AAAAAARILIAO4&data=05%7C01%7Ckadaukes%40chop.edu%7Cb1c7e775ecd44062e36d08dab1428e9e%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017196661126347%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=EBoMjN1pyIHnPjwfqmFBzEtVaH2ryG7TCtO%2FgtooHvI%3D&reserved=0>.
You are receiving this because your review was requested.Message ID: ***@***.***>

** This email originated from an EXTERNAL sender to CHOP. Proceed with caution when replying, opening attachments, or clicking links. Do not disclose your CHOP credentials, employee information, or protected health information to a potential hacker**.

File: R/checks.R (hunk 1 above)
Comment (rsh52) [reply R/checks.R:81]: Sounds good, resolving using callout and change from `all_of` to `any_of`, pushing commit now:

```
  safe_cols <- c(names(db_data)[1], "redcap_event_name",
                 "redcap_repeat_instrument", "redcap_repeat_instance")

  # Step (2)
  check_data <- db_data %>%
    mutate(
      across(.cols = -any_of(safe_cols),
             .names = "{.col}_repeatingcheck",
     ........
```

**Review by ezraporter — APPROVED** [review]
Looks good!

This shouldn't hold this up but we may want to think about whether we want to go all in on `cli` so that our error messages look consistent. Probably better as a separate issue.

File: R/checks.R [hunk 3]
```diff
@@ -57,51 +58,63 @@ check_user_rights <- function(
...
+  #    analysis (i.e. "safe columns").
+  # 2) Create a dummy column for each remaining column and use case_when to
+  #    to assign whether the column demonstrates repeating, nonrepeating, or
+  #    indeterminant behavior per row. Indetemrinant would be the case for
+  #    data not yet filled out.
+  # 3) Use a mapping function to determine if any dummy columns contain both
+  #    "repeating" AND "nonrepeating" declarations, if so error out.
+
+  # Step (1)
+  safe_cols <- c(names(db_data)[1], "redcap_event_name",
```
Comment (skadauke) [inline R/checks.R:82]: Arm?

**Review by skadauke — CHANGES_REQUESTED** [review]

**Comment by skadauke** [comment]
Agreed! Let's make this an issue. Would be a good milestone for 0.2.
________________________________
From: ezraporter ***@***.***>
Sent: Tuesday, October 18, 2022 5:09 PM
To: CHOP-CGTDataOps/REDCapTidieR ***@***.***>
Cc: Stephan Kadauke ***@***.***>; Review requested ***@***.***>
Subject: [External]Re: [CHOP-CGTDataOps/REDCapTidieR] check_repeat_nonrepeat Function Optimization (PR #49)


@ezraporter approved this pull request.

Looks good!

This shouldn't hold this up but we may want to think about whether we want to go all in on cli so that our error messages look consistent. Probably better as a separate issue.

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTDataOps%2FREDCapTidieR%2Fpull%2F49%23pullrequestreview-1146523471&data=05%7C01%7Ckadaukes%40chop.edu%7Cbbeba85d18d44764aca808dab14d04b4%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017241592921587%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=BkwDO7d0CZibPSczNJS3pbmGE5mDqmXSy2m96zRhabU%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTGHWWBEMBWNNCMR22YSXLWD4GXVANCNFSM6AAAAAARILIAO4&data=05%7C01%7Ckadaukes%40chop.edu%7Cbbeba85d18d44764aca808dab14d04b4%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017241593077817%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=i41anB0PdsFi7MXVwdFEf7f9juetB%2BZHQsUDL6BXzdA%3D&reserved=0>.
You are receiving this because your review was requested.Message ID: ***@***.***>

** This email originated from an EXTERNAL sender to CHOP. Proceed with caution when replying, opening attachments, or clicking links. Do not disclose your CHOP credentials, employee information, or protected health information to a potential hacker**.

File: R/checks.R (hunk 3 above)
Comment (rsh52) [reply R/checks.R:82]: Not sure I follow, the standard vars are `redcap_event_name`, `redcap_repeat_instrument`, `redcap_repeat_instance`:
image

We manually split `event_name` into `event` and `arm` later on.

**Comment by skadauke** [comment]
Gotcha! Just checking. I think this is ready to merge!
Thx,
S
________________________________
From: Rich Hanna ***@***.***>
Sent: Tuesday, October 18, 2022 7:30 PM
To: CHOP-CGTDataOps/REDCapTidieR ***@***.***>
Cc: Stephan Kadauke ***@***.***>; Review requested ***@***.***>
Subject: [External]Re: [CHOP-CGTDataOps/REDCapTidieR] check_repeat_nonrepeat Function Optimization (PR #49)


@rsh52 commented on this pull request.

________________________________

In R/checks.R<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTDataOps%2FREDCapTidieR%2Fpull%2F49%23discussion_r998797203&data=05%7C01%7Ckadaukes%40chop.edu%7C4a785b53588c4e8607d008dab160cc14%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017326564952863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=PFLCGJIsfH9tkUQg%2B0WWCraEN56BfxCoVpurFlhitCA%3D&reserved=0>:

> -    repeating <- NA
-
-    for (j in seq_len(nrow(db_data))){
+  # This check function looks for potential repeat/nonrepeat behavior using the
+  # steps below:
+  # 1) Define standard columns that don't need checking and remove those from
+  #    analysis (i.e. "safe columns").
+  # 2) Create a dummy column for each remaining column and use case_when to
+  #    to assign whether the column demonstrates repeating, nonrepeating, or
+  #    indeterminant behavior per row. Indetemrinant would be the case for
+  #    data not yet filled out.
+  # 3) Use a mapping function to determine if any dummy columns contain both
+  #    "repeating" AND "nonrepeating" declarations, if so error out.
+
+  # Step (1)
+  safe_cols <- c(names(db_data)[1], "redcap_event_name",


Not sure I follow, the standard vars are redcap_event_name, redcap_repeat_instrument, redcap_repeat_instance:
[image]<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F38384694%2F196563870-b2962d9d-80f0-4f42-88c2-14faf3ac1247.png&data=05%7C01%7Ckadaukes%40chop.edu%7C4a785b53588c4e8607d008dab160cc14%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017326564952863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=qjkQKMbOFR7%2FUuDwu9Pa7q504shEMkocUienbM%2Fqfdo%3D&reserved=0>

We manually split event_name into event and arm later on.

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTDataOps%2FREDCapTidieR%2Fpull%2F49%23discussion_r998797203&data=05%7C01%7Ckadaukes%40chop.edu%7C4a785b53588c4e8607d008dab160cc14%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017326564952863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=PFLCGJIsfH9tkUQg%2B0WWCraEN56BfxCoVpurFlhitCA%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTGHWSPGW7CI7MLTJZ24CLWD4XKVANCNFSM6AAAAAARILIAO4&data=05%7C01%7Ckadaukes%40chop.edu%7C4a785b53588c4e8607d008dab160cc14%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638017326564952863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=rNANnRfuHIMBOS0wZZNd4YHi6bOWymNY1dtPRcK%2Bn44%3D&reserved=0>.
You are receiving this because your review was requested.Message ID: ***@***.***>

** This email originated from an EXTERNAL sender to CHOP. Proceed with caution when replying, opening attachments, or clicking links. Do not disclose your CHOP credentials, employee information, or protected health information to a potential hacker**.
